src/plugins/data/server/search/es_search/es_search_rxjs_utils.ts

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { Observable } from 'rxjs';
+import type { Logger } from 'kibana/server';
+import type { SearchUsage } from '../collectors';
+import type { IKibanaSearchResponse } from '../../../common/search';
+
+/**
+ * trackSearchStatus is a custom rxjs operator that can be used to track the progress of a search.
+ * @param Logger
+ * @param SearchUsage
+ */
+export const trackSearchStatus = <
+  KibanaResponse extends IKibanaSearchResponse = IKibanaSearchResponse
+>(
+  logger: Logger,
+  usage?: SearchUsage
+) => (source: Observable<KibanaResponse>) =>
+  new Observable<KibanaResponse>((observer) =>
+    source.subscribe({
+      next(response) {
+        const trackSuccessData = response.rawResponse.took;
+
+        if (trackSuccessData) {
+          logger.debug(`trackSearchStatus:next  ${trackSuccessData}`);
+          usage?.trackSuccess(trackSuccessData);
+        }
+        observer.next(response);
+      },
+      error(err) {
+        logger.debug(`trackSearchStatus:error ${err}`);
+        usage?.trackError();
+        observer.error(err);
+      },
+      complete() {
+        observer.complete();
+      },
+    })
+  );
```

Comment:
I think we can simplify this. I was looking at the documentation for custom operators and I think we can do something like this (not I've also added a check for `undefined` since `took` could theoretically be `0`):

```suggestion
export const trackSearchStatus = <
  KibanaResponse extends IKibanaSearchResponse = IEsSearchResponse<SearchResponse<unknown>>
>(
  logger: Logger,
  usage?: SearchUsage
) => {
  return pipe(
    tap(
      (response) => {
        const trackSuccessData = response.rawResponse.took;

        if (trackSuccessData !== undefined) {
          logger.debug(`trackSearchStatus:next  ${trackSuccessData}`);
          usage?.trackSuccess(trackSuccessData);
        }
      },
      (err) => {
        logger.debug(`trackSearchStatus:error ${err}`);
        usage?.trackError();
      }
    )
  );
};
```

I think this will also prevent us from creating a "hot" observable by automatically subscribing to the source in this custom operator.